versioned/spi/src/main/java/org/projectnessie/versioned/LazyPut.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2020 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned;
+
+import java.util.function.Supplier;
+import org.immutables.value.Value;
+import org.projectnessie.model.Content;
+import org.projectnessie.model.ContentKey;
+import org.projectnessie.nessie.relocated.protobuf.ByteString;
+import org.projectnessie.versioned.store.DefaultStoreWorker;
+
+/**
+ * A PUT operation that has been retrieved from the version store. The content value is kept in its
+ * raw (serialized) form; deserialization happens lazily, on first access to {@link #getValue()}.
+ */
+@Value.Immutable
+public interface LazyPut extends Operation {
```

Comment:
Having `Put` and `LazyPut` looks a bit dangerous to me. Both serve the same operation ("put"), but `instanceof` checks against either excludes the other one.

Maybe make `Put` a "bare" interface and have concrete eager and lazy implementations so all users of `Put` can stay as they currently are and no user needs to know what particular implementation will be there.

Potentially also with factory methods only in `Put` (so no top-level type `LazyPut`)

Reply:
So, I managed to reunite `Put` and `LazyPut`. However I had to turn it into an abstract class because I needed a custom equals() method.